core/packages/contracts/src/AgentExecutor.sol

```diff
@@ -0,0 +1,50 @@
+// SPDX-License-Identifier: Apache-2.0
+// SPDX-FileCopyrightText: 2023 Snowfork <hello@snowfork.com>
+pragma solidity 0.8.20;
+
+import {AccessControl} from "openzeppelin/access/AccessControl.sol";
+import {ParaID} from "./Types.sol";
+import {SubstrateTypes} from "./SubstrateTypes.sol";
+
+import {IGateway} from "./IGateway.sol";
+
+import {IERC20Metadata} from "openzeppelin/token/ERC20/extensions/IERC20Metadata.sol";
+import {IERC20} from "openzeppelin/token/ERC20/IERC20.sol";
+import {SafeERC20} from "openzeppelin/token/ERC20/utils/SafeERC20.sol";
+
+import {console} from "forge-std/console.sol";
+
+contract AgentExecutor {
+    using SafeERC20 for IERC20;
+
+    bytes32 public constant ACTION_UNLOCK_TOKENS = keccak256("unlockTokens");
+
+    event NativeTokensUnlocked(address token, address recipient, uint256 amount);
+
+    error InsufficientBalance();
+    error TransferEthFailed();
+    error Unauthorized();
+
+    function execute(address, bytes memory data) external {
```

Comment:
Is it nessessary to check address here only allowed from Gateway?

Reply:
This code is always run via `Agent.invoke`, which does have a check for gateway.

Reply:
Yes that's true. But the `execute` function here is marked as external, without the address check how to differ from an EOA or from an attacker contract?